sdk/dotnet/Pulumi.Automation/InlineProgramArgs.cs

```diff
@@ -0,0 +1,17 @@
+﻿namespace Pulumi.Automation
+{
+    public class InlineProgramArgs : LocalWorkspaceOptions
+    {
+        public string StackName { get; }
+
+        public InlineProgramArgs(
+            string projectName,
+            string stackName,
+            PulumiFn program)
```

Comment:
Yea I made this remark further up about possibly adding async support to `PulumiFn`. There would be trade-offs and in some scenarios the delegate code may be ugly unless we can provide synchronous delegate overloads.

We could make some generic `TStack` overloads in key entrypoints the way the `Deployment.Run` methods do and just pipe the `new TStack();` into our `PulumiFn` delegate to allow for existing `Pulumi.Stack` implementations to be sent in?

Reply:
@mikhailshilkov one thing to think about if we allow the delegate to return a `Task` and as such allow `async` execution is all of the `.InvokeAsync` APIs that return `Task<T>` and are meant to be piped into `Output.Create(Func<Task<T>>)`.

Like If I, instead of sending the `.InvokeAsync` to `Output.Create(...)`, just did `await XXX.InvokeAsync(...);` would that cause any issues or would that work out of the box?

Reply:
That would work if you are in an `async` method. Piping to `Output.Create` is just a way to avoid calling `.Result` inside a sync constructor.

Reply:
OK cool I didn't know if you guys were doing anything special and just hadn't implemented your own TaskAwaiter or something. Well we can go whatever route you guys choose I just chose synchronous to begin with because that is essentially mimicking what we had before doing it in the `Pulumi.Stack` constructor.

Reply:
IMO, it's important to support both `Task`-based functions and `Stack`'s for consistency

Reply:
So, should we have two delegates - a `PulumiFn` and a `PulumiFnAsync`, and just fallback to the async one with overloads for the synchronous one? Or should we just make `PulumiFn` return a task?

Reply:
Two delegates make sense to me

Reply:
OK please checkout the most recent commit and let me know what you think.

I tried to do 2 delegates but it was messy. Because delegates can't do any implicit conversions i required us to have overloads everywhere for sync & async delegates, as well as overloads for `TStack` generic being provided. This was tolerable but still verbose on things like `InlineProgramArgs` where an inline program is required, but it was downright messy on something like `UpOptions` where an inline program was optional - as that required either: 1. sticking with a single async property and requiring the consumer to return a task or 2. having 2 conflicting properties, sync & async, and making sure the consumer only used 1.

Instead I tried changing the `PulumiFn` to a class. This can take in any number of delegates, including a `TStack`. So anywhere we take in `PulumiFn` we now support `TStack`. And it is very easy for us to change the shapes of any delegates that we want to allow as those changes will be encapsulated within `PulumiFn` and won't affect our other APIs since the overloads are no longer necessary.

Reply:
Disregard there is an issue with this that I need to resolve.

Reply:
Alright it is good now and there is a test for `TStack` usage. Some changes to internals had to be made to that `TStack` was passed down to `IRunner` correctly so that the output registration continues to occur as it should in `IRunner`. Let me know how you feel about these changes to `PulumiFn`.

Reply:
I gave this a try and it looks great. I'm happy with what you came up with here!

Reply:
Awesome! I was wondering if you guys like `PulumiFn.Create(...)` or `PulumiFn.New(...)`, I waffled between the two and went with Create but I keep going back and forth.

Reply:
`Create` sounds natural to me 👍